### PR TITLE
Adding virtio support for arm

### LINKED ISSUE #210: Add support for virtio block for aarch64
Configure the virtio block device on arm as well. For this, we need to update the FDT builder to support adding virtio block devices, and update the unit tests such that they work on arm as well.

The definition of done for this task is to:
- make it possible to configure one virtio block
- add tests to validate that this configuration work (we should be able to reuse the existing tests from x86_64)

Out of scope:
- the end to end integration test is out of scope because it requires extensive changes to the S3 bucket & the way we're building the test resources. Any changes that are required once we have a test image should be part of that PR instead.

### PR DESCRIPTION
**Pull request by ramyak-mehra**
Closes #210

Signed-off-by: Ramyak mehra <rmehra_be19@thapar.edu>

### Summary of the PR

*Please summarize here why the changes in this PR are needed.*

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

File: src/arch/src/lib.rs [hunk 1]
```diff
@@ -100,7 +102,13 @@ impl FdtBuilder {
         self.rtc = Some((addr, size));
         self
     }
-
+    pub fn add_virito_device(&mut self, addr: u64, size: u64, irq: u32) -> &mut Self {
```
Comment (andreeaflorescu) [inline src/arch/src/lib.rs:105]: typo: virito instead of virtio.

File: src/arch/src/lib.rs [hunk 2]
```diff
@@ -100,7 +102,13 @@ impl FdtBuilder {
...
+        self.virtio_devices.push((addr, size, irq));
+        self
+    }
```
Comment (andreeaflorescu) [inline src/arch/src/lib.rs:108]: Please add a blank line after each function definition. This is needed for all function you're defining.

File: src/arch/src/lib.rs [hunk 3]
```diff
@@ -319,10 +329,29 @@ fn create_rtc_node(fdt: &mut FdtWriter, rtc_addr: u64, size: u64) -> Result<()>
     Ok(())
 }
 
+fn create_virito_node(fdt: &mut FdtWriter, addr: u64, size: u64, irq: u32) -> Result<()> {
```
Comment (andreeaflorescu) [inline src/arch/src/lib.rs:332]: typo: virito -> virtio.

Same typo on line 333.

File: src/devices/src/virtio/mod.rs [hunk 4]
```diff
@@ -331,11 +336,24 @@ pub(crate) mod tests {
                 kernel_cmdline: &mut self.kernel_cmdline,
             }
         }
+        #[cfg(target_arch = "aarch64")]
+        // Support only GIC_V3.
```
Comment (andreeaflorescu) [inline src/devices/src/virtio/mod.rs:341]: Let's add more comments here: adding a dummy GIC V3 so that we can register the IRQ fds and thus test the virtio functionality in arm as well.

File: src/vmm/src/lib.rs [hunk 5]
```diff
@@ -564,7 +564,7 @@ impl Vmm {
             DEFAULT_ALLOC_POLICY,
         )?;
         self.fdt_builder.with_rtc(range.start(), range.len());
-        let range = mmio_from_range(range);
+        let range = mmio_from_range(&range);
```
Comment (andreeaflorescu) [inline src/vmm/src/lib.rs:567]: We should have this change in a separate commit. If that's too much hassle, we can keep it in this commit but we need to be explicit in the commit message to why this change is needed.

**Review by andreeaflorescu — COMMENTED** [review]
Can you add more details in the commit message? I am expecting to see stuff like:
- a description of what is needed for adding these devices on arm (i.e. the fdt)
- what other changes were required for this to be possible (i.e. the updates on the Mock interface, why were they needed? because otherwise you cannot register_irqfd)
- any changes that you did as part of this commit that are not directly related to the support (i.e. the `mmio_from_range` receives a reference now).

This subtleties are not evident from reading the code and there is no other way of understanding why this commit introduces the changes that it does.

File: src/arch/src/lib.rs [hunk 6]
```diff
@@ -319,10 +338,35 @@ fn create_rtc_node(fdt: &mut FdtWriter, rtc_addr: u64, size: u64) -> Result<()>
...
     use crate::FdtBuilder;
 
+    #[test]
+    fn test_adding_virtio() {
+        let mut fdt = FdtBuilder::new();
+        let fdt = fdt.add_virtio_device(0x1000, 1000, 5);
+        assert_eq!(fdt.virtio_devices.len(), 1);
+    }
+    #[test]
+    fn test_virtio_device_len() {
```
Comment (andreeaflorescu) [inline src/arch/src/lib.rs:365]: This test and the one above seem to be doing the same thing. Can you delete one of them?

**Review by gsserge — APPROVED** [review]

**Review by andreeaflorescu — APPROVED** [review]

**Review by gsserge — APPROVED** [review]
